// @expo/dev-server/build/JsInspector.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.queryAllInspectorAppsAsync = exports.queryInspectorAppAsync = exports.openJsInspector = void 0;
const node_fetch_1 = __importDefault(require("node-fetch"));
const open_1 = __importStar(require("open"));
const path_1 = __importDefault(require("path"));
function openJsInspector(app) {
    // To update devtoolsFrontendRev, find the full commit hash in the url:
    // https://chromium.googlesource.com/chromium/src.git/+log/refs/tags/{CHROME_VERSION}/chrome/VERSION
    //
    // 1. Replace {CHROME_VERSION} with the target chrome version
    // 2. Click the first log item in the webpage
    // 3. The full commit hash is the desired revision
    const devtoolsFrontendRev = 'e3cd97fc771b893b7fd1879196d1215b622c2bed'; // Chrome 90.0.4430.212
    const urlBase = `https://chrome-devtools-frontend.appspot.com/serve_rev/@${devtoolsFrontendRev}/inspector.html`;
    const ws = app.webSocketDebuggerUrl.replace('ws://[::]:', 'localhost:');
    const url = `${urlBase}?panel=sources&v8only=true&ws=${encodeURIComponent(ws)}`;
    launchChromiumAsync(url);
}
exports.openJsInspector = openJsInspector;
async function queryInspectorAppAsync(metroServerOrigin, appId) {
    var _a;
    const apps = await queryAllInspectorAppsAsync(metroServerOrigin);
    return (_a = apps.find(app => app.description === appId)) !== null && _a !== void 0 ? _a : null;
}
exports.queryInspectorAppAsync = queryInspectorAppAsync;
async function queryAllInspectorAppsAsync(metroServerOrigin) {
    const resp = await node_fetch_1.default(`${metroServerOrigin}/json/list`);
    const apps = await resp.json();
    return apps.filter(app => app.vm !== "don't use");
}
exports.queryAllInspectorAppsAsync = queryAllInspectorAppsAsync;
async function launchChromiumAsync(url) {
    // For dev-client connecting metro in LAN, the request to fetch sourcemaps may be blocked by Chromium
    // with insecure-content (https page send xhr for http resource).
    // Adding `--allow-running-insecure-content` to overcome this limitation
    // without users manually allow insecure-content in site settings.
    // However, if there is existing chromium browser process, the argument will not take effect.
    // We also pass a `--user-data-dir=` as temporary profile and force chromium to create new browser process.
    const tmpDir = require('temp-dir');
    const tempProfileDir = path_1.default.join(tmpDir, 'expo-inspector');
    const launchArgs = [
        `--app=${url}`,
        '--allow-running-insecure-content',
        `--user-data-dir=${tempProfileDir}`,
        '--no-first-run',
        '--no-startup-window',
        '--no-default-browser-check',
    ];
    const result = await open_1.openApp(open_1.default.apps.chrome, {
        arguments: launchArgs,
        newInstance: true,
        wait: true,
    });
    if (result.exitCode !== 0) {
        await open_1.openApp(open_1.default.apps.edge, {
            arguments: launchArgs,
            newInstance: true,
            wait: true,
        });
    }
}
